lib/experimental/site-editor-permalinks.php

```diff
@@ -0,0 +1,148 @@
+<?php
+
+add_action(
+	'block_editor_settings_all',
+	function ( $settings ) {
+		$settings['__experimentalDashboardLink'] = admin_url();
+		return $settings;
+	}
+);
+
+function gutenberg_remove_query_args( $args ) {
+	$query_string = $_SERVER['QUERY_STRING'];
+	$query        = wp_parse_args( $query_string );
+	foreach ( $args as $arg_name ) {
+		unset( $query[ $arg_name ] );
+	}
+	return $query;
+}
```

Comment:
Can't we just use [`remove_query_arg`](https://developer.wordpress.org/reference/functions/remove_query_arg/)? It supports multiple query args FWIW.

Reply:
Got it 👍 but 2 things to note here:

- Pretty permalinks don't really work in the admin just yet
- If pretty permalinks are made to work in the admin, I'd expect them to work the same way as on the frontend, meaning, if you get an ugly permalink URL, WP will use the canonical redirect to redirect it to the pretty permalink automatically, meaning, it will still work in the end.

Reply:
> Pretty permalinks don't really work in the admin just yet

They actually work, You can revert this PR to this commit `7cbd26946b1bea2371f08293ab53f6647e11564e` and explore it.

The reason I decided to not do it is that not all of WP installs support mod_rewrite even though it's mentioned in the WP requirements.

Reply:
Interesting 👍 Thanks for mentioning that!

This likely means we either:
- Need to always use the ugly ones (like many plugins do with the REST API for example)
- Have a wrapper that "knows" if pretty permalinks are enabled, and rewrite URLs to it, just like `get_permalink()` or any other canonical link function in WP does it.

Reply:
Anyway, this is a diversion from the original comment. If you just want to remove the query args, you can always do it with simple means like `remove_query_arg( array_keys( $_GET ) )` or something similar - no need for 10 LoC IMHO.

Reply:
The problem is not in the removal of the query args is In the return value. I prefer the current code because it maps properly the way frontend work. A site editor url being a "path" + "query" and the special "p" argument in the query is just an implementation logic because we don't have pretty permalinks. Using `remove_query_arg` means we updated our code to something like 

```
add_query_arg( array('p' => '/navigation' ), remove_query_arg( array( 'postType' ) ) )
```

I agree it's shorter but we lose the "site editor url abstraction (path + query). Maybe it's ok I don't know.

Reply:
About the permalinks:
- Gutenberg can safely call `add_rewrite_rule` without any checks whether `mod_rewrite` is enabled etc. Core does that all the time, like when registering post types or taxonomies or setting up REST API. And so do plugins like WooCommerce
- a common way to detect whether permalinks are enabled is with `get_option( 'permalink_structure' )`. Core functions like `get_trackback_url`, `get_post_embed_url` or `get_rest_url` do this check and return a pretty or ugly URL depending on the result. `gutenberg_get_site_editor_url` could work the same way
- one thing that is special about Site Editor URL is that they don't redirect to the main `index.php` file, but to something else, namely `wp-admin/site-editor.php`. The `WP_Rewrite` class will think that this is an external redirect, i.e., that the rule redirects you outside WordPress. Not sure what are the consequences. But we might want to add a new rule category (`$this->admin_rules`) to `WP_Rewrite`.